modules/standard/IO.chpl

```diff
@@ -3171,7 +3190,34 @@ private proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
   // (it shouldn't release anything since it's a local copy).
   defer { reader._channel_internal = QIO_CHANNEL_PTR_NULL; }
 
-  try x.readThis(reader);
+  if isNilableClassType(t) {
+    // future - write class IDs, have serialization format, handle binary
+    var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
+    var iolit:ioLiteral;
+    if st == QIO_AGGREGATE_FORMAT_JSON {
+      iolit = new ioLiteral("null");
+    } else {
+      iolit = new ioLiteral("nil");
+    }
+    var e:syserr = ENOERR;
+    e = try _read_one_internal(_channel_internal, iokind.dynamic, iolit, loc);
+    if !e {
+      x = nil;
+      return ENOERR;
+    }
+  }
+
+  if isClassType(t) {
+    if x == nil then
+      throw new IllegalArgumentError("cannot read into a nil class");
+
+    var tmp = x!;
+    try tmp.readThis(reader);
+    // the read cannot change the class pointer
+    if tmp != x! then halt ("internal error - class pointer changed");
```

Comment:
This seems similar to the checks above, but for my understanding.. we need this for unmanaged classes, right?

Reply:
It is redundant. It's just that if that property changes then this code will need to change.